### PR TITLE
Optimize prover boundary evaluations in construction of composition polynomial

### PR DESCRIPTION
**Pull request by ajgara**
Improves 2 times the prover speed.

- Computes outside of the forloop some polynomial evaluations and inverses.
- Change lagrange evaluation to FFT.
- Uses batch inverse.

### CODE REVIEW

**Review by Oppen — APPROVED** [review]
LGTM.

File: src/air/constraints/evaluator.rs [hunk 1]
```diff
@@ -83,11 +87,37 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
+            let mut evaluations = evaluate_polynomial_on_lde_domain(
+                poly,
+                domain.blowup_factor,
+                domain.interpolation_domain_size,
+                &domain.coset_offset,
+            )
+            .unwrap();
+            FieldElement::inplace_batch_inverse(&mut evaluations);
+            boundary_zerofiers_inverse_evaluations.push(evaluations);
+        }
```
Comment (Oppen) [inline src/air/constraints/evaluator.rs:114]: We can evaluate the polynomials before collecting the original vectors, so we halve the memory taken by the whole operation.

**Review by diegokingston — APPROVED** [review]

File: src/air/constraints/evaluator.rs (hunk 1 above)
Comment (schouhy) [reply src/air/constraints/evaluator.rs]: Good idea. Done!

**Comment by MauroToscano** [comment]
Good results
Screenshot 2023-06-06 at 14 18 54

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**